### PR TITLE
feat: struct get for non-mut fields is pure

### PR DESCRIPTION
**Pull request by JonathanStarup**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
What about some tests?

**Comment by magnus-madsen** [comment]
@mlutze Do you agree we should have this feature?

File: main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintGen.scala [hunk 1]
```diff
@@ -605,9 +605,10 @@ object ConstraintGen {
         val (instantiatedFieldTpes, structTpe, regionVar) = instantiateStruct(field.sym.structSym, root.structs)
         val (tpe, eff) = visitExp(exp)
         c.expectType(structTpe, tpe, exp.loc)
-        val fieldTpe = instantiatedFieldTpes(field.sym)
+        val (mutable, fieldTpe) = instantiatedFieldTpes(field.sym)
         c.unifyType(fieldTpe, tvar, loc)
-        c.unifyType(Type.mkUnion(eff, regionVar, loc), evar, exp.loc)
+        if (mutable) c.unifyType(Type.mkUnion(eff, regionVar, loc), evar, exp.loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintGen.scala:610]: I think a comment would be appropriate.

**Comment by mlutze** [comment]
> @mlutze Do you agree we should have this feature?

I think our effect system should be as honest as possible

**Comment by JonathanStarup** [comment]
I've tried to make it more clear

I also think of this just as a more precise type rather than a feature

**Comment by magnus-madsen** [comment]
> > @mlutze Do you agree we should have this feature?
> 
> I think our effect system should be as honest as possible

Is  that a yes? Doesnt it depend on what a heap effect is?

**Comment by mlutze** [comment]
> > > @mlutze Do you agree we should have this feature?
> > 
> > 
> > I think our effect system should be as honest as possible
> 
> Is that a yes? Doesnt it depend on what a heap effect is?

It is a yes.

**Comment by JonathanStarup** [comment]
> > > @mlutze Do you agree we should have this feature?
> > 
> > 
> > I think our effect system should be as honest as possible
> 
> Is that a yes? Doesnt it depend on what a heap effect is?

It is an immutable value so It can definitely be pure. The only reason it wouldn't be is if we deallocate the object on scope exit, but I deem that impossible because of algebraic effects

**Comment by magnus-madsen** [comment]
If you add a few tests I would be happy to merge this.

File: main/test/flix/Test.Exp.Struct.Get.flix [hunk 2]
```diff
@@ -78,12 +77,21 @@ mod Get {
...
+        };
+        f("Hello") == "Hello Lucky"
+    }
 }
 
 struct BinaryTree[t, r] {
-    left: Option[BinaryTree[t, r]],
-    right: Option[BinaryTree[t, r]],
-    value: t
+    mut left: Option[BinaryTree[t, r]],
```
Comment (JonathanStarup) [inline main/test/flix/Test.Exp.Struct.Get.flix:92]: I've added `mut` to the structs since its weird only to have immutable fields

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
@JonathanStarup Tests failures

**Pull request merged**